### ISSUE TITLE
Get navi10 (RX 5700 XT) amdgpu working

### ISSUE THREAD

**Issue by jb55** [NixOS/nixpkgs#68809]
Opening this issue to track progress on getting a shiny new amd graphics card working, in case anyone else is thinking about picking up one of these (RX 5700{, XT})

My branch is here: https://github.com/NixOS/nixpkgs/compare/master...jb55:navi10

Progress

- [x] Kernel works
- [x] GPU firmware loads and appears to work.
- [x] [llvm 9](https://gitlab.freedesktop.org/mareko/mesa/commit/594010e366f911581ca0a4471a9d9fa68116514f)
- [x] x11 working (currently just show a black screen)
- [x] kmscube works
- [x] wayland works

**Comment by jb55** [comment]
Got it working, hopefully the above branch helps those in the future.

The step were: 

1. get linux-5.3
2. get mesa-19.2
3. load the gpu firmware (see commits)
4. pass llvm9 to mesa 19.2

I'm going to wait for the non-rc releases before I do a PR.

**Comment by sfultong** [comment]
Thanks for posting your branch! I'm going to try it with my own RX 5700 XT.

Are you going to try to get this into 19.09?

**Comment by jb55** [comment]
Yeah perhaps I'll keep this open until this stuff is actually merged

**Comment by meutraa** [comment]
Yeah, I got this branch working last week. I did have one problem with the xf86-video-amdgpu-git package. It uses fetchgit but imports fetchurl, when I changed it to import fetchgit, and updated the hash, it just did not build, but I got it running with the standard package anyway.

**Comment by sfultong** [comment]
Yeah, it's not building for me. What do you mean, "standard package"?

**Comment by meutraa** [comment]
Reverting this change https://github.com/NixOS/nixpkgs/commit/9c1e7a6cc72acb072dde09e74d7a1686ec0f1769

**Comment by sfultong** [comment]
Ah, thanks. I'll try that for the time being.

**Comment by jb55** [comment]
yeah so `xf86-video-amdgpu-git` was most likely not needed, and `libdrm-git`, but I updated them just in case. the thing that fixed x11 for me was llvm9 for mesa-19.2

**Comment by jb55** [comment]
I guess I forgot to mention that you also need to add the firmware to your configuration:

```
  hardware.firmware = [ pkgs.amdgpu-navi10-firmware ];
```

but this should probably be done automatically when choosing the amdgpu driver

**Comment by sfultong** [comment]
@jb55 ah, yeah, I was still having issues and was wondering if I needed something in my config for the firmware. I'll try that.

**Comment by meutraa** [comment]
LLVM 9, linux 5.3 are both released now, and the navi10 firmware was added to linux-firmware this morning. Now just mesa 19.2 I think :D

**Comment by meutraa** [comment]
Mesa was listening, 19.2 is out :dancing_men:

**Comment by meutraa** [comment]
I believe

4080270cb69bfc62903b5f3fb3eb9ccbe5251ae8 (linux 5.3)
a4340caa4d307f5c27c25d5b60b72378cb345eed (zfs 0.8.2)
987d4fe0717efd398ab88b96735cd3cbb24bffbc (llvm 9)

can be reverted now since master has these versions.

**Comment by meutraa** [comment]
Yeah, I made that change yesterday. I can confirm it removes the need for the navi10 firmware package.

**Comment by meutraa** [comment]
I think that just leaves the firmware-linux-nonfree to be merged and this issue can be closed.

**Comment by jb55** [comment]
I think everything has been merged now... going to test it next time I get the chance.

**Comment by meutraa** [comment]
Just tested master with the following cherry-picked from staging

f22e7f5de004195fb1398ab9605bfc93a7aa6412 (mesa: 19.2.0 -> 19.2.1)
c0c69261dfdfe581f880d4eb3fd0f3aad7ce0715 (SDL2: patch build with mesa 19.2)
53a97c34478e4790ddfa3540d754d87cf423fd81(mesa's llvm: 7 -> 9)

and everything seems to be working fine.

**Comment by jb55** [comment]
@meutraa great! thanks for testing. looks like we're done here

**Comment by meutraa** [comment]
For anyone still following this issue, all the necessary changes are in the master branch as of today.

**Comment by msteen** [comment]
Did these changes make it into `nixos-19.09`? I ask, because I just got a RX 5700 XT and it works with `videoDrivers = [ "vesa" ];` (it reports it is using kernel module `amdgpu`), but the performance is lacking. I don't have configured anything special for it, so maybe I am missing some configuration?

**Comment by jb55** [comment]
what do you mean the performance is lacking? what are you using to test performance?

**Comment by msteen** [comment]
@jb55 Let me first say, thank you for looking into this and making it work! It was one of the reasons I finally decided upon this GPU.

Under `nixos-19.09` if I supply `videoDrivers = [ "amdgpu" ];` it causes it to freeze (no longer even terminal access). If I supply `videoDrivers = [ "vesa" ];` it does work and reports using the `amdgpu` kernel module with `lspci -v`, but when I test the GPU performance by playing the game Factorio, it will be laggy and report a around 15fps. After rebuilding my system with `nixos-unstable` X11 failed to start with `videoDrivers = [ "vesa" ];`, but after changing it to `videoDrivers = [ "amdgpu" ];` it works perfectly and when I now play Factorio I get the same max. fps I see on Windows 10.

So, if it is supposed to work correctly under 19.09, it does not, but you can at least get the GPU somewhat functional (already an achievement considering my past experiences with other GPUs). However under unstable it works as I think you envisioned it, it at least seems to work perfectly so far.

**Comment by jb55** [comment]
fwiw I set `videoDrivers = []` and it works fine. the kernel should be smart enough to load the right driver (amdgpu).

I don't know if it was backported to 19.09, probably not, as there are many changes needed to make this work.

**Comment by Enzime** [comment]
@msteen took a look at both the `release-19.09` and `staging-19.09` branches, and they are missing LLVM 9, mesa 19.2, firmware-linux-nonfree 2019-09-23, so at the current rate it doesn't look like Navi10 will work very well on 19.09.

**Comment by msteen** [comment]
@Enzime Yeah, I indeed noticed the missing LLVM bump to 9, too bad it is not backported, but considering the driver support still seems somewhat unstable, it makes sense for it to still only be available in `nixos-unstable`. So far I have yet to notice any hangup or issue, but I guess it could not hurt to use the latest graphics drivers in the case of such a recent GPU, so thanks for the kernel suggestion!

**Comment by turboMaCk** [comment]
are there any instructions on how to make this work? ` pkgs.amdgpu-navi10-firmware` no longer exists and I had no success with setting `xserver.videoDrivers`. I'm on `nixos-unstable`. So far no matter what I did I always booted to the terminal only.

**Comment by jb55** [comment]
It should just work, as long as you're using the latest linux kernel. In
nixpkgs packageOverrides I have this set:

  linuxPackages = super.linuxPackages_latest;

**Comment by Enzime** [comment]
To run the latest kernel, add to your `configuration.nix`:

    boot.kernelPackages = pkgs.linuxPackages_latest;

The following line is no longer necessary as the Navi10 firmware is included in `firmware-linux-nonfree` package:

    hardware.firmware = [ pkgs.amdgpu-navi10-firmware ];

Instead you should use the following line, which installs `firmware-linux-nonfree`:

    hardware.enableRedistributableFirmware = true;

I'm not entirely sure how necessary setting the `services.xserver.videoDrivers` is, but I do it for good measure:

    services.xserver.videoDrivers = [ "amdgpu" ];

**Comment by turboMaCk** [comment]
Thanks guys! This worked for me. In addition to this I had to **remove** `services.compton.backend = "glx"`(or set it to `services.compton.backend = "xrender" which is also default). Now the the systems seems stabel to me :tada:

**Comment by msteen** [comment]
@Enzime Thanks again for that Linux kernel PR! The GPU would start to glitch when coming out of sleep, but building and using the custom Linux as per your PR seemed to have fixed that.

**Comment by RodrigoLeiteF** [comment]
This doesn't work for me. Using the settings described, my system freezes when starting X11. Using the 5700 XT + Ryzen 5 3600

**Comment by msteen** [comment]
@RodrigoLeiteF Did you upgrade to `nixos-unstable`? It caused freezes for me too under `nixos-19.09`, since the dependencies were not the right versions yet as mentioned above. Other than that, try not setting `services.xserver.videoDrivers = [ "amdgpu" ];` and just keep using the default, it worked for me that way even under `nixos-19.09`, so I could at least use the system. Other than that, the usual applies, check your journal, even if X11 fails, you should still be able to access a terminal with `ctrl+alt+f2` for example.

**Comment by RodrigoLeiteF** [comment]
Upgrading to `nixos-unstable` worked. Thank you! :)

**Comment by akrger** [comment]
> 
> 
> @RodrigoLeiteF Other than that, try not setting `services.xserver.videoDrivers = [ "amdgpu" ];` and just keep using the default, it worked for me that way even under `nixos-19.09`, so I could at least use the system. 

I would advise against that, because for me, setting the `TearFree` option didn't work (got `BadName (named color or font does not exist)`) without setting the driver to `amdgpu` prior.

With that option, I finally have a tear free experience on Linux, with no compositor in between.

**Comment by msteen** [comment]
@akrger As would I, but note I said that in the context of just getting it to work, i.e. to then get it actually properly fixed:
> so I could at least use the system

I totally forgot about that option and was already happy I finally got higher frame rates on my desktop. I just tested it by temporarily enabling it with `xrandr --output HDMI-A-0 --set TearFree on` and I can clearly see the difference between it being on or off, awesome, thanks for the tip!
